hoot-core/src/main/cpp/hoot/core/io/OsmChangesetSqlFileWriter.cpp

```diff
@@ -0,0 +1,420 @@
+    {
+      _createChangeSet();
+      changes = 0;
+    }
+  }
+
+  _outputSql.close();
+}
+
+void OsmChangesetSqlFileWriter::_createChangeSet()
+{
+  if (!_useInternalIds)
+  {
+    _changesetId = _db.getNextId("changesets");
+  }
+  else
+  {
+    _changesetId++;
+  }
+  _outputSql.write(
+    QString("INSERT INTO changesets (id, user_id, created_at, closed_at) VALUES "
+            "(%1, %2, now(), now());\n")
+      .arg(_changesetId)
+      .arg(ConfigOptions().getChangesetUserId())
+    .toUtf8());
+}
+
+long OsmChangesetSqlFileWriter::_getNextId(const ElementType type)
+{
+  switch (type.getEnum())
+  {
+    case ElementType::Node:
+      if (!_useInternalIds)
+      {
+        _nodeId = _db.getNextId("current_" + type.toString().toLower() + "s");
+      }
+      else
+      {
+        _nodeId++;
+      }
+      return _nodeId;
+    case ElementType::Way:
+      if (!_useInternalIds)
+      {
+        _wayId = _db.getNextId("current_" + type.toString().toLower() + "s");
+      }
+      else
+      {
+        _wayId++;
+      }
+      return _wayId;
+    case ElementType::Relation:
+      if (!_useInternalIds)
+      {
+        _relationId = _db.getNextId("current_" + type.toString().toLower() + "s");
+      }
+      else
+      {
+        _relationId++;
+      }
+      return _relationId;
+    default:
+      throw HootException("Unknown element type");
+  }
+}
+
+void OsmChangesetSqlFileWriter::_createNewElement(const ConstElementPtr newElement)
+{
+  switch (newElement->getElementType().getEnum())
+  {
+    case ElementType::Node:
+      _create(dynamic_pointer_cast<const Node>(newElement));
+      break;
+    case ElementType::Way:
+      _create(dynamic_pointer_cast<const Way>(newElement));
+      break;
+    case ElementType::Relation:
+      _create(dynamic_pointer_cast<const Relation>(newElement));
+      break;
+    default:
+      throw HootException("Unknown element type");
+  }
+}
+
+void OsmChangesetSqlFileWriter::_updateExistingElement(const ConstElementPtr updatedElement)
+{
+  switch (updatedElement->getElementType().getEnum())
+  {
+    case ElementType::Node:
+      _modify(dynamic_pointer_cast<const Node>(updatedElement));
+      break;
+    case ElementType::Way:
+      _modify(dynamic_pointer_cast<const Way>(updatedElement));
+      break;
+   case ElementType::Relation:
+      _modify(dynamic_pointer_cast<const Relation>(updatedElement));
+      break;
+    default:
+      throw HootException("Unknown element type");
+  }
+}
+
+void OsmChangesetSqlFileWriter::_deleteExistingElement(const ConstElementPtr removedElement)
+{
+  ElementType::Type elementType = removedElement->getElementType().getEnum();
+  if (elementType != ElementType::Node && elementType != ElementType::Way &&
+      elementType != ElementType::Relation)
+  {
+    throw HootException("Unknown element type");
+  }
+  const QString elementName = removedElement->getElementType().toString().toLower();
+
+  _throwErrorIfElementAtVersionDoesntExist(
```

Comment:
Very good point about the efficiency.  I think an error thrown would be better.

And there is definitely a race condition like you're saying, b/c this would be running inside of a transaction and it wouldn't ever know if anything outside of the transaction updated the element.  So, it would end up writing inconsistent/stale data since there are no server side element version consistency checks in this case like when writing xml to rails port or hoot web services.